helix-rest/src/main/java/org/apache/helix/rest/metadatastore/TrieRoutingData.java

```diff
@@ -0,0 +1,161 @@
+  // some input data. The constructor is blocked by the implementation of RoutingDataAccessor, and
+  // will therefore be implemented later.
+  public TrieRoutingData(TrieNode rootNode) {
+    _rootNode = rootNode;
+  }
+
+  static class TrieNode {
+    final Map<String, TrieNode> _children;
+    final boolean _isLeaf;
+    final String _zkRealmAddress;
+
+    TrieNode(Map<String, TrieNode> children, String name, boolean isLeaf, String zkRealmAddress) {
+      _children = children;
+      _isLeaf = isLeaf;
+      _zkRealmAddress = zkRealmAddress;
+    }
+  }
+
+  public Map<String, String> getAllMappingUnderPath(String zkPath) {
+    TrieNode curNode;
+    try {
+      curNode = findTrieNode(zkPath);
+    } catch (IllegalArgumentException e) {
+      return Collections.emptyMap();
+    }
+
+    Map<String, String> resultMap = new HashMap<>();
+    if (zkPath.substring(zkPath.length() - 1).equals("/")) {
+      zkPath = zkPath.substring(0, zkPath.length() - 1);
+    }
+    addAllAddressesToMapping(resultMap, curNode, zkPath);
+    return resultMap;
+  }
+
+  public String getZkRealm(String zkPath) throws IllegalArgumentException {
+    TrieNode leafNode = findLeafTrieNodeAlongPath(zkPath);
+    return leafNode._zkRealmAddress;
+  }
+
+  /**
+   * Using the root node, find the trie node that the given zkPath is pointing to and return it.
+   * Raise IllegalArgumentException if the zkPath does not point to any node.
+   * @param zkPath - the zkPath where the search is conducted
+   * @return the node pointed by the zkPath
+   * @throws IllegalArgumentException - when the zkPath points to nothing
+   */
+  private TrieNode findTrieNode(String zkPath) throws IllegalArgumentException {
+    if (zkPath.equals("/") || zkPath.equals("")) {
+      return _rootNode;
+    }
+
+    if (zkPath.substring(0, 1).equals("/")) {
+      zkPath = zkPath.substring(1);
+    }
+    String[] splitZkPath = zkPath.split("/", 0);
+    TrieNode curNode = _rootNode;
+    Map<String, TrieNode> curChildren = curNode._children;
+    for (String pathSection : splitZkPath) {
+      curNode = curChildren.get(pathSection);
+      if (curNode != null) {
+        curChildren = curNode._children;
+      } else {
+        throw new IllegalArgumentException("the provided zkPath is missing from the trie");
+      }
+    }
+    return curNode;
+  }
+
+  /**
+   * Using the root node, find the leaf node along the provided zkPath. Raise
+   * IllegalArgumentException if the zkPath does not point to any node or if there is no leaf node
+   * along the path.
+   * @param zkPath - the zkPath where the search is conducted
+   * @return the leaf node along the provided zkPath
+   * @throws IllegalArgumentException - when the zkPath points to nothing or when there is no leaf
+   *           node along the path.
+   */
+  private TrieNode findLeafTrieNodeAlongPath(String zkPath) throws IllegalArgumentException {
```

Comment:
This method is 90% the same as the findTrieNode, can we please merge them?

Reply:
I could set a flag to reuse the same logic, but I was concerned about the code being a bit messy. I'll write it up and see how it turns out.

Reply:
I combined the two methods. It seems a bit long to me, but it does work.